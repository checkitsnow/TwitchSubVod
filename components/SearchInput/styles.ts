import styled from 'styled-components';

interface ContainerProps {
  disabled: boolean;
}

export const Container = styled.form<ContainerProps>`
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;

  form {
    display: flex;
    justify-content: center;
    flex-direction: column;
  }

  > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
  }

  input {
    background: var(--light-background);
    height: 2rem;
    width: 20rem;
    padding: 0;
    color: var(--text);
    text-align: center;
    margin-bottom: 0.5rem;
  }

  button[type='submit'] {
    display: flex;
    align-items: center;
    justify-content: center;
    background: var(--purple);
    height: 2rem;
    width: 20rem;
    border: 2px solid var(--purple);
    color: var(--button-text);
    cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
    transition: all 0.1s ease;

    :hover {
      background: var(--dark-purple);
      border: 2px solid var(--dark-purple);
    }

    & svg {
      margin-right: 0.5rem;
    }
  }
`;
