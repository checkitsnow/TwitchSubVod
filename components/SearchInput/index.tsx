import React, { useEffect, useRef, useState } from 'react';
import ReactGA from 'react-ga';
import { FiSearch } from 'react-icons/fi';
import api from '@/utils/services/api';
import apiV2 from '@/utils/services/apiV2';
import { useGlobal } from '@/stores/GlobalContext';
import { Container } from './styles';
import QualitySelection from '../QualitySelection';
import ErrorModal from '../ErrorModal';
import LoadingModal from '../LoadingModal';

interface AutoCompleteProps {
  display_name: string;
}

interface Props {
  withQuality?: boolean;
}

const SearchInput = ({ withQuality }: Props) => {
  const inputRef = useRef<any>();
  const {
    setVideoQuality,
    error,
    setError,
    loading,
    setLoading,
    setVodUrl,
  } = useGlobal();
  const [username, setUsername] = useState('');
  const [autoComplete, setAutoComplete] = useState<Array<AutoCompleteProps>>(
    [],
  );

  useEffect(() => {
    !!username.length &&
      apiV2
        .get(`search/channels?query=${username}&first=3`)
        .then((response) => {
          setAutoComplete(response.data.data);
        })
        .catch((err) => {
          console.warn(err);
        });
  }, [username]);

  const handleSubmit = () => {
    if (username) {
      setLoading(true);
      try {
        setError('');
        api
          .get(`users?login=${username}`)
          .then((response) => {
            try {
              api
                .get(`channels/${response.data.users[0]._id}/videos?limit=1`)
                .then((channelResponse: any) => {
                  if (channelResponse.data.videos.length !== 0) {
                    location.href = `/videos/${username}`;
                  }
                  if (channelResponse.data._total === 0) {
                    setError(`${username} does not have any available streams`);
                  }
                });

              ReactGA.event({
                category: 'SearchedUserForDeletedVod',
                action: `${username}`,
              });
            } catch (err) {
              setLoading(false);
              setError('This user does not exist or is unavailable');
            }
          })
          .catch((err) => {
            setError(err.response?.data?.message);
            setLoading(false);
          });
      } catch (err) {
        console.warn(err);
        setLoading(false);
        setError('Something went wrong');
      }
    } else {
      setError('Enter a streamer username');
    }
  };

  const isDisabled = username.length <= 0;

  const handleQuality = (event: any) => {
    setVideoQuality(event.target.value);
    setVodUrl('');
  };

  return (
    <Container
      onSubmit={(event) => {
        event.preventDefault();
        handleSubmit();
      }}
      disabled={isDisabled}
    >
      <input
        ref={inputRef}
        type="text"
        name="username"
        aria-label="username"
        onChange={(event) => setUsername(event.target.value)}
        value={username}
        placeholder="Streamer Username"
        list="streamer-username"
        autoComplete="off"
      />

      <datalist id="streamer-username">
        {!!username &&
          !!autoComplete.length &&
          autoComplete.map((user: { display_name: string }) => (
            <option key={user.display_name} value={user.display_name} />
          ))}
      </datalist>
      {withQuality && (
        <QualitySelection onChange={(event: any) => handleQuality(event)} />
      )}
      <button type="submit" aria-label="submit" disabled={isDisabled}>
        <FiSearch size={14} />
        Search
      </button>
      {loading && <LoadingModal />}
      {error && <ErrorModal message={error} />}
    </Container>
  );
};

export default SearchInput;
