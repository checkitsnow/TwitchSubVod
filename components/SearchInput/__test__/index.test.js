import React from 'react';
import renderer from 'react-test-renderer';
import SearchInput from '@/components/SearchInput';

it('should render Search Input correctly', () => {
  const tree = renderer.create(<SearchInput />).toJSON();
  expect(tree).toMatchSnapshot();
});
