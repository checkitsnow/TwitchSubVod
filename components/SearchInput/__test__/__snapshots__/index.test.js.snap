// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Search Input correctly 1`] = `
<form
  className="styles__Container-sc-5265yl-0 lOldp"
  onSubmit={[Function]}
>
  <input
    aria-label="username"
    name="username"
    onChange={[Function]}
    placeholder="Streamer Username"
    type="text"
    value=""
  />
  <div
    className="styles__Container-ezdgk0-0 eWRZyP"
  >
    <label
      htmlFor="quality"
    >
      Select a quality: 
    </label>
    <select
      id="quality"
      name="quality"
      onChange={[Function]}
    >
      <option
        defaultValue="chunked"
      >
        Source
      </option>
      <option
        value="720p60"
      >
        720p60
      </option>
      <option
        value="720p30"
      >
        720p30
      </option>
      <option
        value="480p30"
      >
        480p
      </option>
      <option
        value="360p30"
      >
        360p
      </option>
      <option
        value="160p30"
      >
        160p
      </option>
      <option
        value="audio_only"
      >
        audio
      </option>
    </select>
    <div
      className="styles__Container-pgyzn6-0 iSQhKg"
    >
      <div
        className="info-box"
      >
        <p>
          Works best with "Source"
        </p>
      </div>
      <svg
        fill="none"
        height={18}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <line
          x1="12"
          x2="12"
          y1="16"
          y2="12"
        />
        <line
          x1="12"
          x2="12.01"
          y1="8"
          y2="8"
        />
      </svg>
    </div>
  </div>
  <button
    aria-label="submit"
    type="submit"
  >
    <svg
      fill="none"
      height={14}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width={14}
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="11"
        cy="11"
        r="8"
      />
      <line
        x1="21"
        x2="16.65"
        y1="21"
        y2="16.65"
      />
    </svg>
    Search
  </button>
</form>
`;
